source/Halibut/Transport/ConnectionManagerAsync.cs

```diff
@@ -0,0 +1,300 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Diagnostics;
+using Halibut.Transport.Protocol;
+using Nito.AsyncEx;
+
+namespace Halibut.Transport
+{
+    public class ConnectionManagerAsync : IConnectionManager
+    {
+        readonly ConnectionPoolAsync<ServiceEndPoint, IConnection> pool = new();
+        readonly Dictionary<ServiceEndPoint, HashSet<IConnection>> activeConnections = new();
+
+        // We have separate locks for connections in general (including the pool) vs specifically activeConnections.
+        // This is because disposing calls OnConnectionDisposed, which causes deadlocks if we are not careful.
```

Comment:
I think it would make sense to describe here how to use these locks, when do I take one out over the other.

Reply:
Done